### PR TITLE
Add multi-tenant NetworkContainer controller

### PR DESCRIPTION
**Pull request by feiskyer**
**Reason for Change**:

This PR adds multi-tenant NetworkContainer controller, which includes:

- the CRD reconciler and controller for multi-tenant NetworkContainer
- new interfaces in cnsclient that are consumed in the reconciler
- the mockclients for kubernetes and cns client
- unit tests for the reconciler and controller

It also splits the original `CreateOrUpdateNC()` into two function `CreateOrUpdateNC()` and `UpdateIPAMPoolMonitor()`, so that the new `CreateOrUpdateNC()` interface could be reused across different CRD reconcilers.

**Issue Fixed**:

**Requirements**:

- [x] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [x] includes documentation
- [x] adds unit tests

**Notes**:

### CODE REVIEW

**Review by levimm — APPROVED** [review]

File: cns/multitenantcontroller/multitenantoperator/multitenantcrdcontroller.go [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+	}
+
+	return nil
+}
+
+// return if RequestController is started
+func (mtc *multiTenantController) IsStarted() bool {
+	defer mtc.lock.Unlock()
+	mtc.lock.Lock()
+	return mtc.Started
```
[the remark this answers is not in the document]
Comment (feiskyer) [reply cns/multitenantcontroller/multitenantoperator/multitenantcrdcontroller.go:127]: yes.

File: cns/NetworkContainerContract.go [hunk 2]
```diff
@@ -69,9 +69,10 @@ const (
 
 // ChannelMode :- CNS channel modes
 const (
-	Direct  = "Direct"
-	Managed = "Managed"
-	CRD     = "CRD"
+	Direct         = "Direct"
+	Managed        = "Managed"
+	CRD            = "CRD"
+	MultiTenantCRD = "MultiTenantCRD"
```
[the remark this answers is not in the document]
Comment (feiskyer) [reply cns/NetworkContainerContract.go:75]: it is used in CNS config file.

[the remark this answers is not in the document]
Comment (feiskyer) [reply cns/NetworkContainerContract.go:75]: @thatmattlong DNC should still use CRD, correct ?

[the remark this answers is not in the document]
Comment (thatmattlong) [reply cns/NetworkContainerContract.go:75]: Yes DNC will still use the "CRD" channel

File: cns/multitenantcontroller/multitenantoperator/multitenantcrdreconciler.go [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+			return ctrl.Result{}, err
+		}
+
+		logger.Printf("NC has been terminated for %s", nc.Spec.UUID)
+		return ctrl.Result{}, nil
+	}
+
+	// Do nothing if the network container hasn't been initialized yet from control plane.
+	if nc.Status.State != NCStateInitialized {
+		return ctrl.Result{}, nil
```
Comment (thatmattlong) [inline cns/multitenantcontroller/multitenantoperator/multitenantcrdreconciler.go:76]: Should we return error here so the request will be requeued?

**Comment by thatmattlong** [comment]
Hey @feiskyer there are some merge conflicts in the makefile, after you are able to resolve I will approve

File: cns/multitenantcontroller/multitenantoperator/multitenantcrdreconciler.go (hunk 3 above)
Comment (feiskyer) [reply cns/multitenantcontroller/multitenantoperator/multitenantcrdreconciler.go:76]: No, it should wait for the next update when the state changes.

**Comment by feiskyer** [comment]
@thatmattlong rebased the PR

**Review by thatmattlong — APPROVED** [review]

**Pull request merged**